### PR TITLE
DAKO-96 connect metrics api

### PR DESCRIPTION
**Pull request by ChiaEnC**
## Why do we need this PR?
* connect metrics api to backend

## How did you address the issue?
* implement saga, reducer, and actions 

<br />  

> #### PR readiness check list
> - [x] Did it pass the Flake8/Standard JS check?
> - [x] Did you test this PR?  
> - [x] Did you explain the changes clearly?
> - [x] Does the PR title reference a Jira issue for the correct project (at the beginning of the title)?

### CODE REVIEW

**Review by KalleChen — APPROVED** [review]

File: backend/app/app/crud/crud_metrics.py [hunk 1]
```diff
@@ -15,13 +15,13 @@ def get_growth(
     ) -> Metrics:
         new_customers_this_interval = db.query(Customer).filter(
             Customer.platform_id == metrics_in.platform_id,
-            Customer.created_at <= metrics_in.start_time,
-            Customer.created_at > metrics_in.start_time - timedelta(seconds=metrics_in.interval * 1),
+            Customer.created_at <= datetime.strptime(metrics_in.start_time, "%Y-%m-%d"),
```
Comment (amjltc295) [inline backend/app/app/crud/crud_metrics.py:18]: Use dateutil.parser.parse()

File: website/src/api/fetch.js [hunk 2]
```diff
@@ -89,6 +96,7 @@ export default class Fetcher {
     }
     if (method === 'POST' || method === 'PUT') {
       if (this.contentType === 'application/json') {
+        console.log(typeof obj['start_time']);
```
Comment (amjltc295) [inline website/src/api/fetch.js:99]: Please remove this debugging message

File: website/src/logged_in/views/dashboard/Dashboard.js [hunk 3]
```diff
@@ -1,11 +1,90 @@
...
+    }
+  }
+})
+
+function Dashboard (props) {
+  const { classes, getMetrics, metrics } = props;
+  // TODO : change flexibility 
+  const params = {}
+  params['platform_id'] = 2
+  params['start_time'] = '2019-06-02';
```
Comment (amjltc295) [inline website/src/logged_in/views/dashboard/Dashboard.js:58]: We should set start_time by current time

File: website/src/logged_in/views/dashboard/Dashboard.js [hunk 4]
```diff
@@ -1,11 +1,90 @@
...
+      paddingBottom: theme.spacing(1)
+    }
+  }
+})
+
+function Dashboard (props) {
+  const { classes, getMetrics, metrics } = props;
+  // TODO : change flexibility 
+  const params = {}
+  params['platform_id'] = 2
```
Comment (amjltc295) [inline website/src/logged_in/views/dashboard/Dashboard.js:57]: Why is this platform_id hard-coded?

**Review by amjltc295 — CHANGES_REQUESTED** [review]

File: backend/app/app/schemas/metrics.py [hunk 5]
```diff
@@ -1,22 +1,21 @@
...
     interval: Optional[int] = None  # In second
 
 
 # Properties to receive on metrics creation
 class MetricsGet(MetricsBase):
     platform_id: int
-    start_time: datetime
-    end_time: datetime
+    start_time: str
+    end_time: str
```
Comment (amjltc295) [inline backend/app/app/schemas/metrics.py:18]: If you need to change this type, please add a validation function

File: website/src/sagas/metricsSaga.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+      payload: metrics
+    })
+  } catch (err) {
+    const message = getErrorMessage(err)
+    yield put({
+      type: types.GET_METRICS_ERROR
+    })
+    yield put(setSnackbarErrorAction(message))
+  }
+} else if (error) {
```
Comment (amjltc295) [inline website/src/sagas/metricsSaga.js:41]: Indentation here seems weird

File: backend/app/app/schemas/metrics.py (hunk 5 above)
Comment (amjltc295) [reply backend/app/app/schemas/metrics.py:18]: This is not fixed yet?

**Review by amjltc295 — CHANGES_REQUESTED** [review]

File: backend/app/app/crud/crud_metrics.py [hunk 7]
```diff
@@ -1,9 +1,10 @@
 from datetime import timedelta
-
+from datetime import datetime
 from sqlalchemy.orm import Session
 
 from app.models import Customer, CustomerEvent
 from app.schemas.metrics import Metrics, MetricsGet
+import dateutil.parser
```
Comment (amjltc295) [inline backend/app/app/crud/crud_metrics.py:7]: This is a third-party module; please move it to be with sqlalchemy

File: backend/app/app/crud/crud_metrics.py [hunk 8]
```diff
@@ -1,9 +1,10 @@
 from datetime import timedelta
-
+from datetime import datetime
```
Comment (amjltc295) [inline backend/app/app/crud/crud_metrics.py:2]: Please merge with line 1

File: website/src/sagas/metricsSaga.js (hunk 6 above)
Comment (ChiaEnC) [reply website/src/sagas/metricsSaga.js:41]: do you mean line 40's } ?

**Review by amjltc295 — CHANGES_REQUESTED** [review]

File: backend/app/app/crud/crud_metrics.py [hunk 9]
```diff
@@ -15,13 +16,13 @@ def get_growth(
...
-            Customer.created_at <= metrics_in.start_time,
-            Customer.created_at > metrics_in.start_time - timedelta(seconds=metrics_in.interval * 1),
+            Customer.created_at <= dateutil.parser.parse(metrics_in.start_time),
+            Customer.created_at > dateutil.parser.parse(metrics_in.start_time) - timedelta(seconds=metrics_in.interval * 1),
         ).count()
         new_customers_last_interval = db.query(Customer).filter(
-            Customer.platform_id == metrics_in.platform_id,
-            Customer.created_at <= metrics_in.start_time - timedelta(seconds=metrics_in.interval),
-            Customer.created_at > metrics_in.start_time - timedelta(seconds=metrics_in.interval * 2),
+            Customer.platform_i == metrics_in.platform_id,
```
[the remark this answers is not in the document]
Comment (ChiaEnC) [reply backend/app/app/crud/crud_metrics.py]: I've changed the typo error and pushed ?

File: backend/app/app/crud/crud_metrics.py [hunk 10]
```diff
@@ -1,11 +1,13 @@
 from datetime import timedelta
 
 from sqlalchemy.orm import Session
+import dateutil.parser
```
Comment (amjltc295) [inline backend/app/app/crud/crud_metrics.py:4]: Please remove unnecessary changes

File: website/src/logged_in/views/dashboard/Dashboard.js [hunk 11]
```diff
@@ -1,11 +1,93 @@
...
+      getMetrics({ params })
+    }
+  })
+  console.log(metrics)
 
-function Dashboard () {
   return (
     <div>
-      Dashboard
+      <div className={classes.firstSection}> Interval : {params.interval} ms</div>
```
Comment (amjltc295) [inline website/src/logged_in/views/dashboard/Dashboard.js:70]: It is not ms

File: website/src/logged_in/views/dashboard/Dashboard.js [hunk 12]
```diff
@@ -1,11 +1,93 @@
...
+      <div className={classes.secondSection} />
+      <div className={classes.firstSection}> Platform : {metrics ? metrics.platform_id : null} </div>
+      <div className={classes.secondSection} />
+      <div className={classes.firstSection}> Retention {metrics ? metrics.retention : null} </div>
+      <div className={classes.secondSection} />
+      <div className={classes.firstSection}> Growth {metrics ? metrics.growth : null} </div>
```
Comment (amjltc295) [inline website/src/logged_in/views/dashboard/Dashboard.js:76]: Please include start_time as well

File: website/src/sagas/metricsSaga.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  if (success) {
+    if (store.getState().platform.length === 0) {
+      yield put({
+        type: types.GET_METRICS_ERROR,
+        payload: 'Platform not found'
+      })
+      return
+    }
+    const platformId = store.getState().platform[0].id
+    console.log(platformId)
```
Comment (amjltc295) [inline website/src/sagas/metricsSaga.js:26]: If you want to keep the log here, add "platformID:"

**Review by amjltc295 — CHANGES_REQUESTED** [review]

File: website/src/logged_in/views/dashboard/Dashboard.js (hunk 11 above)
Comment (amjltc295) [reply website/src/logged_in/views/dashboard/Dashboard.js:70]: Should be seconds

File: website/src/sagas/metricsSaga.js (hunk 13 above)
Comment (ChiaEnC) [reply website/src/sagas/metricsSaga.js:26]: deleted ! It's for debugging.

File: backend/app/app/crud/crud_metrics.py [hunk 14]
```diff
@@ -6,6 +6,7 @@
 from app.schemas.metrics import Metrics, MetricsGet
 
 
+
```
Comment (amjltc295) [inline backend/app/app/crud/crud_metrics.py:9]: This too...

**Review by amjltc295 — APPROVED** [review]
LGTM

**Pull request merged**